File: No_105/Construct Binary Tree from Preorder and Inorder Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        TreeNode* root;
        root = solve(0, preorder.size()-1, 0, inorder.size()-1, preorder, inorder);
        return root;
    }
    
    TreeNode* solve(int preStart, int preEnd, int inStart, int inEnd, vector<int>& preorder, vector<int>& inorder) {
        if(preStart > preEnd) return nullptr;
        
        int v = preorder[preStart];
        TreeNode* n = new TreeNode(v);
        
        int count = 0;
        for(int i=inStart; i<=inEnd; i++) {
            if(inorder[i] == v) break;
            count++;
        }
        
        n->left = solve(preStart+1, preStart+count, inStart, inStart+count-1, preorder, inorder);
        n->right = solve(preStart+count+1, preEnd, inStart+count+1, inEnd, preorder, inorder);
            
        return n;
    }
};

File: No_105/Construct Binary Tree from Preorder and Inorder Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        if len(preorder) == 0:
            return None
        
        i = 0
        while preorder[0] != inorder[i]:
            i += 1
            
        root = TreeNode(inorder[i])
        root.left = self.buildTree(preorder[1:1+i], inorder[0:i])
        root.right = self.buildTree(preorder[1+i:], inorder[i+1:])
        
        return root